### ISSUE TITLE
ViewChild is not updated correctly

### ISSUE THREAD

**Issue by danicomas** [angular/angular#13857]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
1) click on a row of the main datatable
2) click on other row 
3) click on the button you will see that the value of the ViewChild is not changed
http://plnkr.co/edit/i0S9e9?p=preview

The element referenced in the ViewChild should contain the current value not the value of the first element obtained.

**Expected behavior**
The element referenced in the ViewChild should contain the current value not the value of the first element obtained.

**Minimal reproduction of the problem with instructions**
1) click on a row of the main datatable
2) click on other row 
3) click on the button you will see that the value of the ViewChild is not changed
http://plnkr.co/edit/i0S9e9?p=preview

**What is the motivation / use case for changing the behavior?**

**Please tell us about your environment:**

* **Angular version:** 2.4.2

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]

* **Node (for AoT issues):** `node --version` =

**Comment by DzmitryShylovich** [comment]
Pls remove all unrelated to angular 3rd party code (ui router, primeng etc). Reproduction should be as simple and small as possible.

**Comment by matsko** [comment]
Yes please do clean up the example a bit. It's hard to follow if it's an issue with your code directly or something within the stack of 3rd party plugins.

**Comment by danicomas** [comment]
@DzmitryShylovich @matsko I have created this plnkr removing the primeng datatable.

To test it you will need to do:
1)Click on "Click Me 1" 
2)Click on "Click Me 2" 
3)Click on button "do". You will see the data of the first element instead than the second row

 http://plnkr.co/edit/hId7wG?preview&p=preview

**Comment by danicomas** [comment]
The plnkr was update @pkozlowski-opensource @DzmitryShylovich @matsko. Some answer?

**Comment by alexzuza** [comment]
If you use dynamic template manipulation you should take care about disposing data like this:

```
@Component({
  selector: "dynamic",
  template: ``
})
export class DynamicComponent {
  @Input() data: any;
  @Input() template: TemplateRef<any>;

  private viewRef: EmbeddedViewRef<any>;

  constructor(public viewContainer: ViewContainerRef) { }

  tryToDispose() {
    if (!this.viewRef) return;
    this.viewContainer.remove(this.viewContainer.indexOf(this.viewRef));
  }

  ngOnChanges() {
    this.tryToDispose();

    if (!this.template) return;

    this.viewRef = this.viewContainer.createEmbeddedView(this.template, {
      '\$implicit': this.data
    });
  }

  ngOnDestroy() {
    this.tryToDispose();
  }
}

```

This way your template will be always have only one projectable node for your `ChildComponent` query

Here's plunker demonstrating this feature https://plnkr.co/edit/V1YQo7VsMJO9Xnjf3Xtz?p=preview

**Comment by danicomas** [comment]
This is a good way @alexzuza

**Comment by danicomas** [comment]
I have updated to the current version and using ng-template and it seems that the problem was solved. https://github.com/angular/angular/issues/13857#issuecomment-271647173
